extra_data/write_cxi.py

```diff
@@ -92,89 +131,96 @@ def _map_chunk(self, chunk, chunk_data, target, tgt_ax1, have_data=None):
             # Look up where the start of this chunk fits in the target
             tgt_start = int(self.train_id_to_ix[chunk_tids[0]])
 
-            target_tids = self.train_ids_perframe[tgt_start : tgt_start+len(chunk_tids)]
-            assert target_tids.shape == chunk_tids.shape, \
-                f"{target_tids.shape} != {chunk_tids.shape}"
+            # Chunk train IDs organized in the same way as in the target
+            chunk_tids_target = np.repeat(chunk_tids, self.ncells)
+
+            target_tids = self.train_ids_perframe[
+                tgt_start : tgt_start + len(chunk_tids)*self.ncells
+            ]
+            assert target_tids.shape == chunk_tids_target.shape, \
+                f"{target_tids.shape} != {chunk_tids_target.shape}"
             assert target_tids[0] == chunk_tids[0], \
                 f"{target_tids[0]} != {chunk_tids[0]}"
 
             # How much of this chunk can be mapped in one go?
-            mismatches = (chunk_tids != target_tids).nonzero()[0]
+            mismatches = (chunk_tids_target != target_tids).nonzero()[0]
             if mismatches.size > 0:
-                n_match = mismatches[0]
+                n_match = int(mismatches[0] / self.ncells)
             else:
                 n_match = len(chunk_tids)
 
             # Select the matching data and add it to the target
             chunk_match_end = chunk_match_start + n_match
+            tgt_end = tgt_start + (n_match*self.ncells)
             matched = chunk_data[chunk_match_start:chunk_match_end]
-            target[tgt_start : tgt_start+n_match, tgt_ax1] = matched
+            if self.ncells == 1:
+                # In some cases, there's an extra dimension of length 1.
+                # E.g. JUNGFRAU data with 1 memory cell per train or
+                # AGIPD/LPD raw data.
+                # Also, 'VirtualSource' object has no attribute 'ndim'.
+                if (len(matched.shape) > 1 and matched.shape[1] == 1):
+                    target[tgt_start:tgt_end, tgt_ax1] = matched[:, 0]
```

Comment:
Unfortunately, I don't think this will quite work correctly with VirtualSource objects. The selection mechanism just applies one selection, so doing a second selection like this will replace the previous selection rather than combining with it.

I think this would only come up if someone tried to assemble a virtual CXI file from LPD or DSSC raw data, so maybe it's not worth worrying about. But if not, maybe we should fail up front in those cases rather than trying something that's broken.

Reply:
Yes, thank you for a very valuable comment and suggestions on improving VDS readout performance, implemented in https://github.com/European-XFEL/EXtra-data/commit/6582e9954511eb0bc752851d8f64a31bb7ed1348.